src/charm.py

```diff
@@ -0,0 +1,595 @@
+
+    Args:
+        func: Action function to be decorated.
+
+    Returns:
+        Decorated charm function with catching unexpected errors.
+    """
+
+    @functools.wraps(func)
+    def func_with_catch_unexpected_errors(self, event: ActionEvent) -> None:
+        # Safe guard against unexpected error.
+        try:
+            func(self, event)
+        except Exception as err:  # pylint: disable=broad-exception-caught
+            logger.exception(err)
+            event.fail(f"Failed to get runner info: {err}")
+
+    return func_with_catch_unexpected_errors
+
+
+class GithubRunnerCharm(CharmBase):
+    """Charm for managing GitHub self-hosted runners."""
+
+    _stored = StoredState()
+
+    service_token_path = Path("service_token")
+    repo_check_web_service_path = Path("/home/ubuntu/repo_policy_compliance_service")
+    repo_check_web_service_script = Path("src/repo_policy_compliance_service.py")
+    repo_check_systemd_service = Path("/etc/systemd/system/repo-policy-compliance.service")
+
+    def __init__(self, *args, **kargs) -> None:
+        """Construct the charm.
+
+        Args:
+            args: List of arguments to be passed to the `CharmBase` class.
+            kargs: List of keyword arguments to be passed to the `CharmBase`
+                class.
+        """
+        super().__init__(*args, **kargs)
+
+        self._event_timer = EventTimer(self.unit.name)
+
+        self._stored.set_default(
+            path=self.config["path"],  # for detecting changes
+            runner_bin_url=None,
+        )
+
+        self.proxies: ProxySetting = {}
+        if http_proxy := get_env_var("JUJU_CHARM_HTTP_PROXY"):
+            self.proxies["http"] = http_proxy
+        if https_proxy := get_env_var("JUJU_CHARM_HTTPS_PROXY"):
+            self.proxies["https"] = https_proxy
+        if no_proxy := get_env_var("JUJU_CHARM_NO_PROXY"):
+            self.proxies["no_proxy"] = no_proxy
+
+        self.service_token = None
+
+        self.on.define_event("reconcile_runners", ReconcileRunnersEvent)
+        self.on.define_event("update_runner_bin", UpdateRunnerBinEvent)
+
+        self.framework.observe(self.on.install, self._on_install)
+        self.framework.observe(self.on.upgrade_charm, self._on_upgrade_charm)
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(self.on.reconcile_runners, self._on_reconcile_runners)
+        self.framework.observe(self.on.update_runner_bin, self._on_update_runner_bin)
+        self.framework.observe(self.on.stop, self._on_stop)
+
+        self.framework.observe(self.on.check_runners_action, self._on_check_runners_action)
+        self.framework.observe(self.on.reconcile_runners_action, self._on_reconcile_runners_action)
+        self.framework.observe(self.on.flush_runners_action, self._on_flush_runners_action)
+        self.framework.observe(self.on.update_runner_bin_action, self._on_update_runner_bin)
```

Comment:
Should this have a final `_action` postfix?

Reply:
It's sharing the same hook as "update_runner_bin" event
Each event should have a dedicated hook and eventually call the same private method

Reply:
Minor comments